File: implementations/dart/lib/src/zobrist.dart
import 'piece.dart';
import 'board.dart';

class Zobrist {
  static final Zobrist instance = Zobrist._();

  final List<List<int>> pieces = List.generate(12, (_) => List.filled(64, 0));
  late final int sideToMove;
  final List<int> castling = List.filled(4, 0);
  final List<int> enPassant = List.filled(8, 0);

  Zobrist._() {
    int state = 0x123456789ABCDEF0;
    const mask64 = 0xFFFFFFFFFFFFFFFF;

    int nextRand() {
      state ^= (state << 13) & mask64;
      state ^= (state >> 7);
      state ^= (state << 17) & mask64;
      return state;
    }

    for (int p = 0; p < 12; p++) {
      for (int s = 0; s < 64; s++) {
        pieces[p][s] = nextRand();
      }
    }

    sideToMove = nextRand();

    for (int i = 0; i < 4; i++) {
      castling[i] = nextRand();
    }

    for (int i = 0; i < 8; i++) {
      enPassant[i] = nextRand();
    }
  }

  int getPieceIndex(Piece piece) {
    int idx = 0;
    switch (piece.type) {
      case PieceType.pawn: idx = 0; break;
      case PieceType.knight: idx = 1; break;
      case PieceType.bishop: idx = 2; break;
      case PieceType.rook: idx = 3; break;
      case PieceType.queen: idx = 4; break;
      case PieceType.king: idx = 5; break;
    }
    if (piece.color == PieceColor.black) {
      idx += 6;
    }
    return idx;
  }

  int computeHash(Board board) {
    int hashVal = 0;
    for (int row = 0; row < 8; row++) {
      for (int col = 0; col < 8; col++) {
        final piece = board.squares[row][col];
        if (piece != null) {
          final square = row * 8 + col;
          final idx = getPieceIndex(piece);
          hashVal ^= pieces[idx][square];
        }
      }
    }

    if (board.turn == 'b') {
      hashVal ^= sideToMove;
    }

    final rights = board.get_castlingRights_internal();
    if (rights.contains('K')) hashVal ^= castling[0];
    if (rights.contains('Q')) hashVal ^= castling[1];
    if (rights.contains('k')) hashVal ^= castling[2];
    if (rights.contains('q')) hashVal ^= castling[3];

    final epTarget = board.get_enPassantTarget_internal();
    if (epTarget != null) {
      hashVal ^= enPassant[epTarget.col];
    }

    return hashVal;
  }
}

File: implementations/php/lib/Zobrist.php
<?php

namespace Chess;

require_once __DIR__ . '/constants.php';

class Zobrist {
    public array $pieces;
    public \GMP $side_to_move;
    public array $castling;
    public array $en_passant;

    private static ?Zobrist $instance = null;

    public static function getInstance(): Zobrist {
        if (self::$instance === null) {
            self::$instance = new Zobrist();
        }
        return self::$instance;
    }

    private function __construct() {
        $this->pieces = array_fill(0, 12, array_fill(0, 64, null));
        $this->castling = array_fill(0, 4, null);
        $this->en_passant = array_fill(0, 8, null);

        $state = gmp_init("0x123456789ABCDEF0");
        $mask64 = gmp_init("0xFFFFFFFFFFFFFFFF");

        $next_rand = function() use (&$state, $mask64) {
            // state ^= state << 13
            $state = gmp_xor($state, gmp_and(gmp_mul($state, gmp_pow(2, 13)), $mask64));
            // state ^= state >> 7
            $state = gmp_xor($state, gmp_div($state, gmp_pow(2, 7)));
            // state ^= state << 17
            $state = gmp_xor($state, gmp_and(gmp_mul($state, gmp_pow(2, 17)), $mask64));
            return $state;
        };

        for ($p = 0; $p < 12; $p++) {
            for ($s = 0; $s < 64; $s++) {
                $this->pieces[$p][$s] = $next_rand();
            }
        }

        $this->side_to_move = $next_rand();

        for ($i = 0; $i < 4; $i++) {
            $this->castling[$i] = $next_rand();
        }

        for ($i = 0; $i < 8; $i++) {
            $this->en_passant[$i] = $next_rand();
        }
    }

    public function get_piece_index(int $type, int $color): int {
        $idx = match($type) {
            CHESS_PAWN => 0,
            CHESS_KNIGHT => 1,
            CHESS_BISHOP => 2,
            CHESS_ROOK => 3,
            CHESS_QUEEN => 4,
            CHESS_KING => 5,
            default => 0
        };
        if ($color === CHESS_BLACK) {
            $idx += 6;
        }
        return $idx;
    }

    public function compute_hash($board): \GMP {
        $hash = gmp_init(0);
        
        for ($row = 0; $row < 8; $row++) {
            for ($col = 0; $col < 8; $col++) {
                [$piece, $color] = $board->squares[$row][$col];
                if ($piece !== CHESS_EMPTY) {
                    $square = (7 - $row) * 8 + $col; // a1=0
                    $idx = $this->get_piece_index($piece, $color);
                    $hash = gmp_xor($hash, $this->pieces[$idx][$square]);
                }
            }
        }

        if ($board->current_player === CHESS_BLACK) {
            $hash = gmp_xor($hash, $this->side_to_move);
        }

        $rights = $board->castling_rights;
        if ($rights->white_kingside) $hash = gmp_xor($hash, $this->castling[0]);
        if ($rights->white_queenside) $hash = gmp_xor($hash, $this->castling[1]);
        if ($rights->black_kingside) $hash = gmp_xor($hash, $this->castling[2]);
        if ($rights->black_queenside) $hash = gmp_xor($hash, $this->castling[3]);

        $ep = $board->en_passant_target;
        if ($ep !== null) {
            $hash = gmp_xor($hash, $this->en_passant[$ep[1]]);
        }

        return $hash;
    }
}
